spec.bs

```diff
@@ -1998,11 +1996,11 @@ dictionary AdRender {
 dictionary GenerateBidOutput {
   required double bid;
   DOMString bidCurrency;
-  required (DOMString or AdRender) render;
+  (DOMString or AdRender) render;
   any ad;
   sequence<(DOMString or AdRender)> adComponents;
   double adCost;
-  double modelingSignals;
+  unrestricted double modelingSignals;

```

Comment:
Just to be clear, this is needed because you want to not throw on inf/-inf and NaN values (i.e., accept them quietly and ignore?). Would it be easier to just make this an unsigned short, matching https://wicg.github.io/turtledove/#generated-bid-modeling-signals, and properly throwing for the invalid values?

Reply:
Well, the explainer explicitly says we don't throw for them, but ignore them. Looks like Caleb added this here:
https://github.com/WICG/turtledove/commit/02dda4b6e2286a49c11b4a58c4ce19244d005a46
... though that doesn't explain why it was decided that way.

It's sort of logical since we don't throw for other things outside [0, 4096),  and, well, I am pretty sure +infinity is outside the range. The weirdness for using unsigned short is that it would throw for 66 000 but not for 5000. That seems weird. If WebIDL had a uint12_t, that would be another matter, but sensibly enough it doesn't.

... The implicit conversion to uint16 in both the spec and the impl isn't great, though. Since this doesn't deal with negatives there are no surprises there, though.

Reply:
Hmm I see. So you don't throw for things outside of [0, 4096) range, but you don't always ignore them right? You ignore negative values, infinity, and NaN, but you do accept and fully use the lower 12 bits of 5000? Is that right?

Reply:
Oh boy. So the impl (and current spec language) completely ignores stuff outside the range, while the explainer has this confusing language about lower 12 bits. (Which would probably suggest an unsigned long long field and a bitmask?)

Reply:
What does the spec and implementation do for 4000.125?

Reply:
The spec: not sure. What happens when you put a 4000.12 in a WebIDL double into infra unsigned short?
... I think it probably needs an explicit floor() here: 
https://github.com/WICG/turtledove/blob/cdc7f80746ff06606fbfb3da51faf106ecc09d4e/spec.bs#L2130

Impl: same as for 4000

Reply:
> The spec: not sure. What happens when you put a 4000.12 in a WebIDL double into infra unsigned short?

Well when converting from an ECMAScript value to an unsigned short we'd go through https://webidl.spec.whatwg.org/#abstract-opdef-converttoint, which does all of the integer conversion and flooring for us etc., so maybe we actually don't need to explicitly floor here? Maybe we can just say:

> Set |bid|'s [=generated bid/modeling signals=] to the result of [=converting=] the ECMAScript value represented by |modelingSignals| to an unsigned short

How about that? That way we're basically converting the unrestricted double directly to an unsigned short as explicitly as possible.

Reply:
Done